allennlp/data/dataset_readers/copynet_seq2seq.py

```diff
@@ -0,0 +1,187 @@
+import logging
+from typing import List, Dict, Tuple, Optional
+
+import numpy as np
+from overrides import overrides
+
+from allennlp.common.checks import ConfigurationError
+from allennlp.common.file_utils import cached_path
+from allennlp.common.util import START_SYMBOL, END_SYMBOL
+from allennlp.data.dataset_readers.dataset_reader import DatasetReader
+from allennlp.data.fields import TextField, ArrayField, MetadataField, CopyMapField
+from allennlp.data.instance import Instance
+from allennlp.data.tokenizers import Token, Tokenizer, WordTokenizer
+from allennlp.data.token_indexers import TokenIndexer, SingleIdTokenIndexer
+
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+@DatasetReader.register("copynet_seq2seq")
+class CopyNetDatasetReader(DatasetReader):
+    """
+    Read a tsv file containing paired sequences, and create a dataset suitable for a
+    ``CopyNet`` model, or any model with a matching API.
+
+    Expected format for each input line: <source_sequence_string><tab><target_sequence_string>.
+
+    `START_SYMBOL` and `END_SYMBOL` tokens are added to the source and target sequences.
+
+    Parameters
+    ----------
+    target_namespace : ``str``, required
+        The vocab namespace for the targets. This needs to be passed to the dataset reader
+        in order to construct the CopyMapField.
+    source_tokenizer : ``Tokenizer``, optional
+        Tokenizer to use to split the input sequences into words or other kinds of tokens. Defaults
+        to ``WordTokenizer()``.
+    target_tokenizer : ``Tokenizer``, optional
+        Tokenizer to use to split the output sequences (during training) into words or other kinds
+        of tokens. Defaults to ``source_tokenizer``.
+    source_token_indexers : ``Dict[str, TokenIndexer]``, optional
+        Indexers used to define input (source side) token representations. Defaults to
+        ``{"tokens": SingleIdTokenIndexer()}``.
+    target_token_indexers : ``Dict[str, TokenIndexer]``, optional
+        Indexers used to define output (target side) token representations. Defaults to
+        ``source_token_indexers``.
```

Comment:
Actually, you don't have a `TextFieldEmbedder` for the targets at all.  You should just remove this argument entirely and use a `ListField[LabelField]` instead of a `TextField` for the targets.

Reply:
Hmm ok so I just added a commit that ensures `self._target_token_indexers` is always just `{"tokens": SingleidTokenIndexer}`. I do get what you're saying with using `ListField[LabelField]` instead, but I have been thinking about making the target embedding more flexible in the future, probably through using `TextFieldEmbedder`, so I think it might be worth it to keep it as a `TextField` for now. What do you think?

Reply:
Yeah, fair enough, if you anticipate making the target embedding more complex, then it's reasonable to leave it as a `TextField`.  Up to you.